### PR TITLE
No More Hunting (seeds edition)

### PR DESCRIPTION
**Pull request by faddat**
- Add x/incentives cli docs (#332)
- Fix CLI bug with start-time being blank
- Fork in BeginBlock code (#395)
- Add Validator Commission upgrade logic
- Go mod tidy
- Add script to be used for testing the upgrade
- Change SDK version to v0.42.9
- Change upgrade height to 712,000 and upgrade to cosmos SDK v0.42.9
- Update testing script
- Add more instructions to testing script
- Add required beginblocker for SDK v0.42.9
- No more pass around document.  Validators add peers and seeds here.  It is then visible, versioned, and requires no thought from new users.  Genesis is next.
- Ended a pointless, harmful, centralization prone tradition.

### CODE REVIEW

**Comment by faddat** [comment]
Basically:

* brings main current with v3.1.0
* Adds default seeds and peers.  We should do this.  Current state of affairs wastes hours of people's lives when moving towards most common setup of osmosis.  
* Next PR in the easy series embeds the genesis.json.  Starting osmosis mainnet will be done as:

```
osmosisd start
```

File: cmd/osmosisd/cmd/init.go
Comment (Sdpierret) [inline cmd/osmosisd/cmd/init.go:84]: peers and seed can change over time.
I think it's better to keep this blank and use external documentation.

Comment (faddat) [reply cmd/osmosisd/cmd/init.go:84]: Sir, that's the point.  They will change over time.  

They will change here, and be versioned, and there will be some sense of accountability and we won't have to pass around a semi-secret, hard to locate document.

Comment (faddat) [reply cmd/osmosisd/cmd/init.go:84]: please consider the harm done by asking 100+ people to routinely configure seeds and peers.  

This also allows us to begin working on formally tracking sync speed:

https://github.com/osmosis-labs/osmosis/pull/418

....and shouldn't we live in a world, where each validator has at least one commit on this repository?

Comment (Sdpierret) [reply cmd/osmosisd/cmd/init.go:84]: it's not something who need to be versioned.
it's a production context.

Comment (faddat) [reply cmd/osmosisd/cmd/init.go:84]: 2 things:

https://twitter.com/gadikian/status/1427959398857543687

Screen Shot 2021-08-18 at 6 51 55 PM

^ thought we had a team just shy of 100?

**Comment by ValarDragon** [comment]
Thanks for adding this!

File: cmd/osmosisd/cmd/init.go
Comment (ValarDragon) [inline cmd/osmosisd/cmd/init.go:86]: +1

File: go.mod
Comment (ValarDragon) [inline go.mod:3]: Why change this?

File: cmd/osmosisd/cmd/init.go
Comment (ValarDragon) [inline cmd/osmosisd/cmd/init.go:84]: +1 to default seeds list. I think we should not be adding default persistent peers though

Comment (ValarDragon) [reply cmd/osmosisd/cmd/init.go:84]: The idea should be to make it easy to get access to nodes in the network, which is what the default seeds should be accomplishing.

Persistent peers makes you keep on trying to get a connection with them. They will get their inbound peers saturated, and then the node will get confused with logs saying that they keep getting refusals to connect. (And it will overly hammer bandwidth of these listed persistent peers)

File: go.mod
Comment (faddat) [reply go.mod:3]: I changed it today because of the new release, and my desire to use `embed`.

File: cmd/osmosisd/cmd/init.go
Comment (faddat) [reply cmd/osmosisd/cmd/init.go:84]: hmmmmmm good point!

We will need more people to run seeds in that case.  

I'll ask in the validator channel.

**Comment by faddat** [comment]
THe seed I put there is running the new tenderseed, which happens to osmo by default:

https://github.com/notional-labs/tenderseed which is of course a fork of a fork

File: go.mod
Comment (ValarDragon) [reply go.mod:3]: What is `embed`? Also `go.1.17` is imo too recent, I definitely think we should be staying on 15/16 for awhile, unless quite important.

File: go.mod
Comment (ValarDragon) [inline go.mod:28]: Why did a second require section get added to the go.mod?

File: cmd/osmosisd/cmd/init.go
Comment (ValarDragon) [inline cmd/osmosisd/cmd/init.go:84]: After talking to folks on Interchain Gmbh & Informal teams, I'm a bit worried about adding a default list of seeds for the following two reasons:

1) If you connect to a seed node, that itself doesn't consider itself to be a seed node, then you will remain connected to that node as a peer (not disconnect). So suppose we have N seeds listed, and `k` were malicious. As long as one of the seeds isn't malicious, I will get peers honestly on the network. However, in the current Tendermint, if those seeds are maliciously misconfigured to not disconnect from you, then they will fill up all your node connections. Then you won't sample any of the other peers you got from the honest seeds, and will just be connected to those `k` malicious seeds, thereby ruining your security.

2) Having a default seed list is great, as long as Tendermint itself is making random sampling guarantees. E.g. if I list 100 seeds, it should do one of:
 -  Get peers from each seed, build a large list, then randomly sample from the large list. 
 - Randomly sample which seeds to query, then build a sufficiently sized peer list. Then randomly sample peers of the local address book to go ping.

Comment (ValarDragon) [reply cmd/osmosisd/cmd/init.go:84]: I view (1) as a bug that needs to be fixed before I feel comfortable with this. 

I view (2) as something that I need to look into to double check, and if whats implemented is one of those scenarios, then lets go with this!

Comment (ValarDragon) [reply cmd/osmosisd/cmd/init.go:84]: Granted, there is an argument to be made that anything done with many validators putting things into this repo is much better than the currently mutable-by-anyone google sheets & hackmd's get passed around.

Thoughts @sunnya97 @UnityChaos ?

Comment (ValarDragon) [reply cmd/osmosisd/cmd/init.go:84]: Actually, I think (1) is fine by ensuring that we only seed nodes who we know the team for. This is similar to what bitcoin does: https://en.bitcoin.it/wiki/Satoshi_Client_Node_Discovery (Only 6 core devs curate lists) Thanks @sunnya97 for bringing this up.

We should still get that bug fixed tho.

(2) is something we should ensure, but less concerning.

**Review by ValarDragon — COMMENTED** [review]
LGTM except for go mod situation. 

For the seed nodes, lets add some comments explaining who are the default seed nodes. Perhaps this can be done by defining a list of seed strings, each with a comment after it saying whose it is. And then combining this list to get the TM format

File: go.mod
Comment (faddat) [reply go.mod:28]: go 1.17 does this.  First I'd seen it, too.

File: cmd/osmosisd/cmd/init.go
Comment (faddat) [reply cmd/osmosisd/cmd/init.go:84]: Yeah, so that is exactly what I am thinking:

we should know the operators.

**Comment by faddat** [comment]
@ValarDragon go mod situation is just normal go 1.17, afaik.

https://golang.org/doc/go1.17#go-command


```
If a module specifies go 1.17 or higher, the module graph includes only the immediate dependencies of other go 1.17 modules, not their full transitive dependencies. (See Module graph pruning for more detail.)

For the go command to correctly resolve transitive imports using the pruned module graph, the go.mod file for each module needs to include more detail about the transitive dependencies relevant to that module. If a module specifies go 1.17 or higher in its go.mod file, its go.mod file now contains an explicit require directive for every module that provides a transitively-imported package. (In previous versions, the go.mod file typically only included explicit requirements for directly-imported packages.) 
```

In general, my thinking is that we should aggressively keep latest versions for go, etc.

wdyt?

**Comment by faddat** [comment]
* go 1.16
* No persistent peers
* 150 inbound peers
* Seeds as a slice with a comment saying who is what by each seed

File: cmd/osmosisd/cmd/init.go
Comment (ValarDragon) [inline cmd/osmosisd/cmd/init.go:98]: I think the default should be `info`?

Comment (faddat) [reply cmd/osmosisd/cmd/init.go:98]: oop, thank you!  I had been testing something.

**Review by ValarDragon — APPROVED** [review]
This PR is good to merge imo, except for the default log level being changed to error. 

Unless theres a reason you think this is a better default, in which case I'd love to discuss! I have a preference for discussing in a second PR tho, so we can unblock this one / get it merged in. 

I currently feel like the info level logging of consensus, where you see blocks fly by, is quite helpful.

File: cmd/osmosisd/cmd/init.go
Comment (UnityChaos) [reply cmd/osmosisd/cmd/init.go]: Yeah, I think well known validator operators making PRs to add their seed nodes would be fine. :+1:

**Comment by ValarDragon** [comment]
I'll fix the go mod issue, just made a branch to fix it last night, and hadn't PR'd it to main yet.

**Comment by faddat** [comment]
ooookay.  I may reopen this anew.  "works on my laptop not good enough."

**Comment by ValarDragon** [comment]
Accidentally squash merged, fixing the commit history in #421

**Comment by ValarDragon** [comment]
Fixed! Ok now the code changes should be trackable

**Pull request merged**
